cinder/volume/drivers/vmware/vmdk.py

```diff
@@ -2236,10 +2369,166 @@ def do_setup(self, context):
+            back = volume_utils.extract_host(volume.host)
+            backend = back.split('@')[1]
+
+            new_host = '{}@{}#{}'.format(
+                host, backend,
+                model['vmware_ds_name']
+            )
+            return new_host
+
+    @utils.trace
+    def update_provider_info(self, volumes, snapshots):
+        """Ensure we have a provider_id set on volumes.
+
+        If there is a provider_id already set then skip, if it is missing then
+        we will update it based on the volume object. We can always compute
+        the id if we have the full volume object, but not all driver API's
+        give us that info.
+
+        We have to save each volume entry if they update their host, otherwise
+        cinder volume manager doesn't see that volume as part of the host
+        allocated_capacity calculation, which happens right after this call
+        completes.
+
+        We don't care about snapshots, they just use the volume's provider_id.
+        """
+        LOG.info("HOST {} : volumes {}".format(self.host, len(volumes)))
+        if self.configuration.vmware_datastores_as_pools:
+            LOG.info("vmware_datastores_as_pools is enabled. "
+                     "Checking host entries for volumes and snapshots.")
+            datastores = self._get_datastores_for_profiles()
+            ds_keys = datastores.keys()
+            vol_updates = []
+            LOG.info("Process {} volumes".format(len(volumes)))
+            for vol in volumes:
+                # make sure we have the correc host info
+                if vol['status'] in ['in-use', 'available']:
+                    # do we need to update the host?
+                    pool_info = volume_utils.extract_host(
+                        vol.host, level='pool', default_pool_name=True)
+
+                    # IF the pool has already been set correctly, then
+                    # no need to make api calls to vcenter to fetch the
+                    # datastore name from the volume backing information.
+                    # this will save time on every startup
+                    if (pool_info not in ds_keys or
+                            pool_info == volume_utils.DEFAULT_POOL_NAME):
+                        LOG.debug("Updating host for volume {}".format(vol.id))
+
+                        try:
+                            new_host = self._new_host_for_volume(vol)
+                            if new_host:
+                                vol.update({'host': new_host})
+                                vol.save()
+                        except Exception as ex:
+                            LOG.warning("Couldn't update host for {} because "
+                                        " {}".format(vol.id, ex))
+                    else:
+                        LOG.debug("Keeping host for volume {}".format(vol.id))
+
+            LOG.info("HOST COMPLETE {}".format(self.host))
+            return vol_updates, None
```

Comment:
We never add anything to `vol_updates`. Is that by design?

Also, `vol_updates` defaults to an empty list, while we return `None` for the snapshot updates. Can whatever calls us cope with `None` or is there something that could try to iterate over the return-value?

Reply:
Yah it's by design now.  When I first wrote this it was sending volume updates back to the volume manager to update each volume, but that caused other problems with the host being still incorrect.  So, I had to update the volume update (and DB) before returning.   I can just force pass back None, None instead of creating a vol_updates now.